aggregator/src/config.rs

```diff
@@ -0,0 +1,329 @@
+    /// Input the hash input bytes,
+    /// assign the circuit for hash function,
+    /// return cells for the hash inputs and digests.
+    #[allow(clippy::type_complexity)]
+    pub(crate) fn assign(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        challenges: Challenges<Value<F>>,
+        preimages: &[Vec<u8>],
+    ) -> Result<
+        (
+            Vec<Vec<AssignedCell<F, F>>>, // input cells
+            Vec<Vec<AssignedCell<F, F>>>, // digest cells
+        ),
+        Error,
+    > {
+        let mut is_first_time = true;
+        let num_rows = 1 << LOG_DEGREE;
+        let witness = multi_keccak(preimages, challenges, capacity(num_rows))?;
+
+        // extract the indices of the rows for which the preimage and the digest cells lie in
+        let (preimage_indices, digest_indices) = get_indices(preimages);
+
+        let mut hash_input_cells = vec![];
+        let mut hash_output_cells = vec![];
+
+        layouter.assign_region(
+            || "assign keccak rows",
+            |mut region| {
+                if is_first_time {
+                    is_first_time = false;
+                    let offset = witness.len() - 1;
+                    self.keccak_circuit_config
+                        .set_row(&mut region, offset, &witness[offset])?;
+                    return Ok(());
+                }
+                // ====================================================
+                // Step 1. Extract the hash cells
+                // ====================================================
+                let mut current_hash_input_cells = vec![];
+                let mut current_hash_output_cells = vec![];
+
+                for (offset, keccak_row) in witness.iter().enumerate() {
+                    let row =
+                        self.keccak_circuit_config
+                            .set_row(&mut region, offset, keccak_row)?;
+
+                    if preimage_indices.contains(&offset) {
+                        current_hash_input_cells.push(row[6].clone());
+                    }
+                    if digest_indices.contains(&offset) {
+                        current_hash_output_cells.push(row.last().unwrap().clone());
+                    }
+
+                    // we reset the current hash when it is finalized
+                    // note that length == 0 indicate that the hash is a padding
+                    // so we simply skip it
+                    if keccak_row.is_final && keccak_row.length != 0 {
+                        hash_input_cells.push(current_hash_input_cells);
+                        hash_output_cells.push(current_hash_output_cells);
+                        current_hash_input_cells = vec![];
+                        current_hash_output_cells = vec![];
+                    }
+                }
+
+                // sanity: we have same number of hash input and output
+                let hash_num = hash_input_cells.len();
+                let num_chunks = hash_num - 2;
```

Comment:
Just to confirm: this -2 comes from the fact we take out `batch_public_input_hash_preimage` and `batch_data_hash_preimage`, and the rest are chunk hash preimages?

Reply:
yes that is correct